src/rgw/rgw_d3n_datacache.h

```diff
@@ -0,0 +1,346 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#ifndef CEPH_RGWD3NDATACACHE_H
+#define CEPH_RGWD3NDATACACHE_H
+
+#include "rgw_rados.h"
+#include <curl/curl.h>
+#include <experimental/filesystem>
+
+#include "rgw_common.h"
+
+#include <unistd.h>
+#include <signal.h>
+#include "include/Context.h"
+#include "include/lru.h"
+#include "rgw_d3n_cacherequest.h"
+
+
+/*D3nDataCache*/
+namespace efs = std::experimental::filesystem;
+struct D3nDataCache;
+
+struct D3nChunkDataInfo : public LRUObject {
+	CephContext *cct;
+	uint64_t size;
+	time_t access_time;
+	string address;
+	string oid;
+	bool complete;
+	struct D3nChunkDataInfo* lru_prev;
+	struct D3nChunkDataInfo* lru_next;
+
+	D3nChunkDataInfo(): size(0) {}
+
+	void set_ctx(CephContext *_cct) {
+		cct = _cct;
+	}
+
+	void dump(Formatter *f) const;
+	static void generate_test_instances(list<D3nChunkDataInfo*>& o);
+};
+
+struct D3nCacheAioWriteRequest{
+	string oid;
+	void *data;
+	int fd;
+	struct aiocb *cb;
+	D3nDataCache *priv_data;
+	CephContext *cct;
+
+	D3nCacheAioWriteRequest(CephContext *_cct) : cct(_cct) {}
+	int create_io(bufferlist& bl, unsigned int len, string oid, string cache_location);
+
+  ~D3nCacheAioWriteRequest() {
+    ::close(fd);
+		cb->aio_buf = nullptr;
+		free(data);
+		data = nullptr;
+		delete(cb);
+  }
+};
+
+struct D3nDataCache {
+
+private:
+  std::map<string, D3nChunkDataInfo*> cache_map;
+  std::list<string> outstanding_write_list;
+  int index;
+  std::mutex cache_lock;
+  std::mutex eviction_lock;
+  std::string cache_location;
+
+  CephContext *cct;
+  enum class _io_type {
+    SYNC_IO = 1,
+    ASYNC_IO = 2,
+    SEND_FILE = 3
+  } io_type;
+  enum class _eviction_policy {
+    LRU=0, RANDOM=1
+  } eviction_policy;
+
+  struct sigaction action;
+  uint64_t free_data_cache_size = 0;
+  uint64_t outstanding_write_size = 0;
+  struct D3nChunkDataInfo* head;
+  struct D3nChunkDataInfo* tail;
+
+private:
+  void add_io();
+
+public:
+  D3nDataCache();
+  ~D3nDataCache() {}
+
+  bool get(const string& oid, const off_t len);
+  void put(bufferlist& bl, unsigned int len, string& obj_key);
+  int io_write(bufferlist& bl, unsigned int len, std::string oid);
+  int create_aio_write_request(bufferlist& bl, unsigned int len, std::string oid);
+  void cache_aio_write_completion_cb(D3nCacheAioWriteRequest* c);
+  size_t random_eviction();
+  size_t lru_eviction();
+
+  void init(CephContext *_cct) {
+    cct = _cct;
+    free_data_cache_size = cct->_conf->rgw_d3n_l1_datacache_size;
+    head = nullptr;
+    tail = nullptr;
+    cache_location = cct->_conf->rgw_d3n_l1_datacache_persistent_path;
+    if(cache_location.back() != '/') {
+       cache_location += "/";
+    }
+    try {
+      if (efs::exists(cache_location)) {
+        // evict the cache storage directory
+        if (g_conf()->rgw_d3n_l1_evict_cache_on_start) {
```

Comment:
i worry about the interaction between `rgw_d3n_l1_datacache_size` and the default `rgw_d3n_l1_evict_cache_on_start = false`. if we don't clean the directory on startup, we lose control over its total size. after some number of radosgw restarts, we're likely to fill the disk and get ENOSPC

i feel like we either need to require evict-on-start, or scan the whole directory on startup to initialize the cache and `free_data_cache_size`. what do you think?

Reply:
corrected (changed to true)

Reply:
> i feel like we either need to require evict-on-start, or scan the whole directory on startup to initialize the cache and free_data_cache_size. what do you think?

if we're not going to implement the scanning on startup, i don't think `rgw_d3n_l1_evict_cache_on_start` should be an option